pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/io/stream/ReadableStreamingData.java

```diff
@@ -130,8 +132,97 @@ public long skip(final long n) {
-        } catch (IOException e) {
+        } catch (final IOException e) {
+            throw new DataAccessException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long readBytes(@NonNull final byte[] dst, final int offset, final int maxLength) {
+        if (maxLength < 0) {
+            throw new IllegalArgumentException("Negative maxLength not allowed");
+        }
+        if (!hasRemaining()) {
+            return 0;
+        }
+        try {
+            int bytesRead = in.read(dst, offset, maxLength);
+            position += bytesRead;
+            if (bytesRead < maxLength) {
+                final int nextByte = tryReadNextByte();
+                if (nextByte != -1) {
+                    dst[offset + bytesRead] = (byte) nextByte;
+                    bytesRead++;
+                }
+            }
+            return bytesRead;
+        } catch (final IOException e) {
             throw new DataAccessException(e);
         }
     }
+
+    @Override
+    public long readBytes(@NonNull final ByteBuffer dst) {
+        if (!dst.hasArray()) {
+            return ReadableSequentialData.super.readBytes(dst);
+        }
+        if (!hasRemaining()) {
+            return 0;
+        }
+        final byte[] dstArr = dst.array();
+        final int dstArrOffset = dst.arrayOffset();
+        final int dstPos = dst.position();
+        final long len = Math.min(remaining(), dst.remaining());
+        try {
+            int bytesRead = in.read(dstArr, dstPos + dstArrOffset, Math.toIntExact(len));
+            position += bytesRead;
+            if (bytesRead < len) {
+                final int nextByte = tryReadNextByte();
+                if (nextByte != -1) {
+                    dst.put((byte) nextByte);
+                    bytesRead++;
+                }
+            }
+            return bytesRead;
+        } catch (final IOException e) {
+            throw new DataAccessException(e);
+        }
+    }
+
+    @Override
+    public long readBytes(@NonNull final BufferedData dst) {
+        final long len = Math.min(remaining(), dst.remaining());
+        int bytesRead = dst.writeBytes(in, Math.toIntExact(len));
+        position += bytesRead;
+        if (position < len) {
+            final int nextByte = tryReadNextByte();
+            if (nextByte != -1) {
+                dst.writeByte((byte) nextByte);
+                bytesRead++;
+            }
+        }
+        return bytesRead;
+    }
+
+    // This method is used when less than requested bytes have been read from the input stream.
+    // It could happen for two reasons: end of stream or because no more bytes are immediately
+    // available. A simple check for in.available() can't be used to detect end of stream. The
+    // workaround is to try to read the next byte and check the result. If it is -1, then the
+    // stream has really reached the end, otherwise this method returns the read byte
+    private byte tryReadNextByte() {
```

Comment:
Rather than this workaround, we should use `readNBytes` if the intent is to block until all requested data is read or we reach end of stream.

Reply:
I was thinking about `readNBytes()`, too, but decided not to proceed with it. Here is a scenario: a request to read N bytes from an input stream, which is known to be much larger than N, but less than N bytes are immediately available. According to `InputStream.read()` spec, it may not read all requested N bytes, but only available bytes. Subsequent reads may read more bytes, as they become available.

PBJ interfaces like `ReadableSequentialData` try to be at the same level as `InputStream` and `ByteBuffer`. This is why semantics of `ReadableSequentialData.readBytes()` is "read as much as available and return the number of bytes read", but not "read exactly the specified number of bytes, or throw an exception" (similar to `readNBytes()`)

In MerkleDb code, we have utility methods `MerkleDbFileUtils.completelyRead()/completelyWrite()`, which are to help with such incomplete reads/writes. Similar methods will be provided at PBJ protobuf level (e.g. in `ProtoParserTools` and `ProtoWriterTools`). IO interfaces like `ReadableSequentialData` are too low-level for that, in my opinion.

@jasperpotts What was the original idea behind PBJ data I/O library APIs? Should they be as low-level as `InputStream` and `ByteBuffer`, or should work as higher level utility methods like `MerkleDbFileUtils.completelyRead()`?

Reply:
readNBytes doesn't throw an exception if it reads too few bytes, it just keeps calling for bytes until it gets the requested number or reaches end of stream.  It's doing _exactly_ what you're doing here, just built into InputStream.

Using readNBytes doesn't change the semantics of your PR, it just reduces the code we have to implement and maintain.

Reply:
@artemananiev , while the PBJ interfaces may try to be at the same level as `InputStream` and `ByteBuffer`, they also have their own specification defined in their JavaDocs. And if I read the JavaDoc for `long ReadableSequentialData.readBytes(@NonNull final byte[] dst, final int offset, final int maxLength)` correctly, then it does indeed align with that of the `InputStream.readNBytes`. So if we can take advantage of this method directly and also simplify our code here, then I agree with @jsync-swirlds . Please let me know if I'm missing any differences in the respective specifications that would invalidate the point.

Reply:
OK, I think I know where most confusion comes from. In JDK terms, "remaining" means the bytes that can be are immediately available, they can be read without blocking. In PBJ, "remaining" semantics is slightly different, it means everything up to the limit.

With that in mind, if `readBytes()` is guaranteed to read `Math.min(remaining, maxLength)` bytes, regardless of how many bytes are immediately available in the stream, then I agree using `InputStream.readNBytes()` is the way to go. Let me make this change. It would simplify the code quite a bit and would eliminate need in this tricky "read one next byte" workaround.

Reply:
This odd method is dropped in favor of using `readNBytes()`